rubix-core/src/main/java/com/qubole/rubix/core/DummyModeCachingInputStream.java

```diff
@@ -0,0 +1,95 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License. See accompanying LICENSE file.
+ */
+package com.qubole.rubix.core;
+
+import com.google.common.base.Throwables;
+import com.qubole.rubix.spi.BookKeeperFactory;
+import com.qubole.rubix.spi.ClusterType;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.thrift.shaded.transport.TTransportException;
+
+import java.io.IOException;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkState;
+
+public class DummyModeCachingInputStream extends CachingInputStream
+{
+  private static final Log log = LogFactory.getLog(DummyModeCachingInputStream.class);
+
+  public DummyModeCachingInputStream(FileSystem parentFs,
+                                     Path backendPath,
+                                     Configuration conf,
+                                     CachingFileSystemStats statsMbean,
+                                     ClusterType clusterType,
+                                     BookKeeperFactory bookKeeperFactory,
+                                     FileSystem remoteFileSystem,
+                                     int bufferSize,
+                                     FileSystem.Statistics statistics)
+          throws IOException
+  {
+    super(parentFs, backendPath, conf, statsMbean, clusterType, bookKeeperFactory, remoteFileSystem, bufferSize, statistics);
+  }
+
+  @Override
+  public long getPos()
+          throws IOException
+  {
+    return getParentDataInputStream().getPos();
+  }
+
+  @Override
+  public void seek(long pos)
+          throws IOException
+  {
+    checkState(pos >= 0, "Negative Position");
+    long currentPos = getPos();
+    log.debug(String.format("Dummy Mode: Seek request, currentPos: %d currentBlock: %d", currentPos, currentPos / blockSize));
+    getParentDataInputStream().seek(pos);
+    log.debug(String.format("Dummy Mode: Seek to %d, setting block location %d", currentPos, currentPos / blockSize));
+  }
+
+  @Override
+  public int read(final byte[] buffer, final int offset, final int length)
+          throws IOException
+  {
+    final long initPos = getPos();
+    final long initNextReadBlock = initPos / blockSize;
+    final int read = readFullyDirect(buffer, offset, length);
+
+    readService.execute(new Runnable()
+    {
+      @Override
+      public void run()
+      {
+        try {
+          long endBlock = ((initPos + (length - 1)) / blockSize) + 1;
+          final List<ReadRequestChain> readRequestChains = setupReadRequestChains(buffer, offset, endBlock, length, initPos, initNextReadBlock, bookKeeperFactory.createBookKeeperClient(conf));
+          updateCacheAndStats(readRequestChains);
```

Comment:
`updateCacheAndStats` will submit a work on readService. We dont need this redirection, the current work is happening on readService thread so let the updation also happen there. You can change `updateCacheAndStats` to do only the actual work, line299 to line305, and let the callers call it in a executor thread

Reply:
Okay will do